### PR TITLE
Add javascript code snippet validation

### LINKED ISSUE #594: cwltool --validate should validate that javascript expressions are valid ES5 expressions
It would be nice if `cwltool --validate` would report ES5 syntax errors in cwl expressions e.g. running `cwltool --validate` with this:
```yaml
cwlVersion: v1.0
requirements:
- class: InlineJavascriptRequirement
inputs:
  - id: parameter
    type: string
    inputBinding:
      valueFrom: $((_ => "test")())
outputs: []
class: CommandLineTool
baseCommand: echo
```
would report an error, as it uses ES6 syntax.

### PR DESCRIPTION
**Pull request by ThomasHickman**
This fixes #594. Note: I refractored most of `exec_js` (in sandbox_js)  into `exec_js_process`, so that I could use a function that runs a javascript process, inputs the results to stdin and captures them on stdout, while having a timeout.

### CODE REVIEW

**Comment by mr-c** [comment]
`cwltool/linting/lint.js` needs to be listed in MANIFEST.in

File: cwltool/validate_js.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    returncode, stdout, stderr = exec_js_process(
+        path.join(linter_folder, "jshint_wrapper.js"),
+        json.dumps({
+            "code": js_text,
+            "options": options,
+            "globals": globals
+        })
+    )
+
+    def dump_jshint_error():
```
Comment (mr-c) [inline cwltool/validate_js.py]: This needs type information

File: cwltool/validate_js.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+            elif validate_ex(possible_schema, tool, raise_ex=False):
+                valid_schema = possible_schema
+
+        return get_expressions(tool, valid_schema, source_line)
+    elif isinstance(schema, avro.schema.ArraySchema):
+        if not isinstance(tool, list):
+            return []
+
+        return list(itertools.chain(*
+            map(lambda x: get_expressions(x[1], schema.items, SourceLine(tool, x[0])), enumerate(tool))
```
Comment (mr-c) [inline cwltool/validate_js.py]: `cwltool/validate_js.py:47: error: "Schema" has no attribute "items"` from https://travis-ci.org/common-workflow-language/cwltool/jobs/348306873#L618

File: cwltool/load_tool.py [hunk 3]
```diff
@@ -187,7 +188,9 @@ def validate_document(document_loader,  # type: Loader
                       fetcher_constructor=None,  # type: FetcherConstructorType
                       skip_schemas=None,  # type: bool
                       overrides=None,  # type: List[Dict]
-                      metadata=None,  # type: Optional[Dict]
+                      metadata=None,  # type: Optional[Dict],
```
Comment (mr-c) [inline cwltool/load_tool.py]: The extra comma at the end of the type hint needs to be removed

File: cwltool/validate_js.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+                "W104", "W119"  # using ES6 features
+            ],
+            "strict": "implied",
+            "esversion": 5
+        }
+
+    linter_folder = resource_filename(__name__, "jshint")
+
+    returncode, stdout, stderr = exec_js_process(
+        path.join(linter_folder, "jshint_wrapper.js"),
```
Comment (mr-c) [inline cwltool/validate_js.py]: Here you should use get_data from https://github.com/common-workflow-language/cwltool/blob/master/tests/util.py (maybe that should be moved to the cwltool package?)

This is why the conformance tests are failing 🙂

**Comment by mr-c** [comment]
@ThomasHickman You are getting closer! :-)

**Comment by ThomasHickman** [comment]
@mr-c any idea what I'm still doing wrong? :)

**Comment by mr-c** [comment]
@ThomasHickman I haven't had a chance to try myself yet, but the next stage is to create a new virtualenv, `pip install` this branch into it; and then execute the CWL conformance tests, debugging any failures directly.

https://github.com/common-workflow-language/common-workflow-language/blob/master/CONFORMANCE_TESTS.md#usage with `RUNNER=cwltool`

**Comment by mr-c** [comment]
Much closer!

```
~/common-workflow-language/v1.0$ cwltool --debug  v1.0/template-tool.cwl v1.0/cat-job.json
/home/michael/common-workflow-language/env/bin/cwltool 1.0.20180310142018
Resolved 'v1.0/template-tool.cwl' to 'file:///home/michael/common-workflow-language/v1.0/v1.0/template-tool.cwl'
I'm sorry, I couldn't load this CWL file.  The error was:
Traceback (most recent call last):
  File "/home/michael/common-workflow-language/env/local/lib/python2.7/site-packages/cwltool/main.py", line 492, in main
    validate_js_options=validate_js_options)
  File "/home/michael/common-workflow-language/env/local/lib/python2.7/site-packages/cwltool/load_tool.py", line 288, in validate_document
    validate_js_expressions(processobj, avsc_names.names[workflowobj["class"]], validate_js_options)
  File "/home/michael/common-workflow-language/env/local/lib/python2.7/site-packages/cwltool/validate_js.py", line 170, in validate_js_expressions
    expression_lib_errors, expression_lib_globals = jshint_js("\n".join(expression_lib), default_globals, jshint_options)
  File "/home/michael/common-workflow-language/env/local/lib/python2.7/site-packages/cwltool/validate_js.py", line 112, in jshint_js
    context=jshint_functions_text
  File "/home/michael/common-workflow-language/env/local/lib/python2.7/site-packages/cwltool/sandboxjs.py", line 264, in exec_js_process
    while not process_finished() and tm.is_alive():
  File "/home/michael/common-workflow-language/env/local/lib/python2.7/site-packages/cwltool/sandboxjs.py", line 186, in process_finished
    return stdout_buf.getvalue().decode().endswith(PROCESS_FINISHED_STR) and \
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 28449: ordinal not in range(128)
```

**Comment by mr-c** [comment]
@ThomasHickman 
This is looking good! I ran `cwltool --validate` from this branch on every CWL in the V1.0 specification directory, for fun.

```
/home/michael/.local/bin/cwltool 1.0.20180310142018
Resolved 'v1.0/template-tool.cwl' to 'file:///home/michael/common-workflow-language/v1.0/v1.0/template-tool.cwl'
v1.0/template-tool.cwl:9:5: JSHINT:       exports = module.exports = _;
v1.0/template-tool.cwl:9:5: JSHINT:       ^
v1.0/template-tool.cwl:9:5: JSHINT: W117: 'exports' is not defined.
v1.0/template-tool.cwl:9:5: JSHINT:     exports._ = _;
v1.0/template-tool.cwl:9:5: JSHINT:     ^
v1.0/template-tool.cwl:9:5: JSHINT: W117: 'exports' is not defined.
v1.0/template-tool.cwl:9:5: JSHINT:     if (typeof module !== 'undefined' && module.exports) {
v1.0/template-tool.cwl:9:5: JSHINT:                                          ^
v1.0/template-tool.cwl:9:5: JSHINT: W117: 'module' is not defined.
v1.0/template-tool.cwl:9:5: JSHINT:       exports = module.exports = _;
v1.0/template-tool.cwl:9:5: JSHINT:                 ^
v1.0/template-tool.cwl:9:5: JSHINT: W117: 'module' is not defined.
v1.0/template-tool.cwl:9:5: JSHINT:     return setTimeout(function(){
v1.0/template-tool.cwl:9:5: JSHINT:            ^
v1.0/template-tool.cwl:9:5: JSHINT: W117: 'setTimeout' is not defined.
v1.0/template-tool.cwl:9:5: JSHINT:         timeout = setTimeout(later, remaining);
v1.0/template-tool.cwl:9:5: JSHINT:                   ^
v1.0/template-tool.cwl:9:5: JSHINT: W117: 'setTimeout' is not defined.
v1.0/template-tool.cwl:9:5: JSHINT:         timeout = setTimeout(later, wait - last);
v1.0/template-tool.cwl:9:5: JSHINT:                   ^
v1.0/template-tool.cwl:9:5: JSHINT: W117: 'setTimeout' is not defined.
v1.0/template-tool.cwl:9:5: JSHINT:       if (!timeout) timeout = setTimeout(later, wait);
v1.0/template-tool.cwl:9:5: JSHINT:                               ^
v1.0/template-tool.cwl:9:5: JSHINT: W117: 'setTimeout' is not defined.
v1.0/template-tool.cwl:9:5: JSHINT:         clearTimeout(timeout);
v1.0/template-tool.cwl:9:5: JSHINT:         ^
v1.0/template-tool.cwl:9:5: JSHINT: W117: 'clearTimeout' is not defined.
v1.0/template-tool.cwl:9:5: JSHINT:     root._ = previousUnderscore;
v1.0/template-tool.cwl:9:5: JSHINT:              ^
v1.0/template-tool.cwl:9:5: JSHINT: W117: 'previousUnderscore' is not defined.
v1.0/template-tool.cwl:9:5: JSHINT:   if (typeof define === 'function' && define.amd) {
v1.0/template-tool.cwl:9:5: JSHINT:                                       ^
v1.0/template-tool.cwl:9:5: JSHINT: W117: 'define' is not defined.
v1.0/template-tool.cwl:9:5: JSHINT:     define('underscore', [], function() {
v1.0/template-tool.cwl:9:5: JSHINT:     ^
v1.0/template-tool.cwl:9:5: JSHINT: W117: 'define' is not defined.
v1.0/template-tool.cwl:9:5: JSHINT: var t = function(s) { return _.template(s, {variable: 'data'})({'inputs': inputs}); };
v1.0/template-tool.cwl:9:5: JSHINT:                              ^
v1.0/template-tool.cwl:9:5: JSHINT: W117: '_' is not defined.
Tool definition is valid
```
This is due to `common-workflow-language/v1.0/v1.0/underscore.js` which isn't mentioned anywhere in the reporting; can that be fixed?
Should this have passed validation?

**Comment by mr-c** [comment]
Jenkins, test this please

**Comment by mr-c** [comment]
Jenkins, test this please

**Comment by ThomasHickman** [comment]
> This is due to common-workflow-language/v1.0/v1.0/underscore.js which isn't mentioned anywhere in the reporting; can that be fixed?
Should this have passed validation?

These errors are generated because: jshint is not taking in to account `if(typeof ... === "undefined")` type statements (issue for this is jshint: https://github.com/jshint/jshint/issues/1986), underscore is using `setTimeout` and `clearTimeout`, which are not in the ES5.1 spec and jshint is not fully understanding the statement `root._ = _` [see here](https://github.com/common-workflow-language/common-workflow-language/blob/83650a5b6b34c0648a20742eb02b475865555f50/v1.0/v1.0/underscore.js#L51) as creating a new variable `_`. This could be fixed to not output errors by: a) using `--disable-js-validation` b) disabling `W117` (undefined name checking), by using `--js-hint-options-file` for anything which uses `underscore.js`. I enabled the `W117`, as it is very useful for detecting errors in simple JavaScript expressions, but for a piece of code which is a large and dynamic as underscore.js, warnings are probably going to come up

**Comment by mr-c** [comment]
Thanks for the explanation. Is there any way to print the actual filename causing the warnings?

**Comment by ThomasHickman** [comment]
^ I spotted an bug in recording the position of the incorrect in the expressionLib. However, as far as I can see, there is no easy way of reporting the file which actually contains the JavaScript which is evaluated (if this is inserted using schema salad)

File: cwltool/validate_js.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    include_warnings = [
+        "W117", # <VARIABLE> not defined
+        "W104", "W119"  # using ES6 features
+    ]
+
+    return error_code[0] == "E" or error_code in include_warnings
+
+JSHintJSReturn = namedtuple("jshint_return", ["errors", "globals"])
+
+def jshint_js(js_text, globals=None, options=None):
```
Comment (mr-c) [inline cwltool/validate_js.py]: this shadows the built in function globals

File: cwltool/validate_js.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    default_globals = [u"self", u"inputs", u"runtime", u"console"]
+
+    for i, prop in enumerate(reversed(requirements)):
+        if prop["class"] == "InlineJavascriptRequirement":
+            expression_lib = prop.get("expressionLib", [])
+            break
+    else:
+        return
+
+    globals = copy.deepcopy(default_globals)
```
Comment (mr-c) [inline cwltool/validate_js.py]: This shadows the built in function globals. Also, copy.deepcopy isn't imported

**Comment by mr-c** [comment]
@ThomasHickman I don't mind the warnings, but I'll defer to you which ones to enable

File: cwltool/sandboxjs.py [hunk 7]
```diff
@@ -131,15 +123,39 @@ def new_js_proc(force_docker_pull=False, js_console=False):
...
-    if not hasattr(localdata, "proc") or localdata.proc.poll() is not None or onWindows():
-        localdata.proc = new_js_proc(force_docker_pull=force_docker_pull, js_console=js_console)
+    if localdata.procs.get(js_engine) is None \
+            or localdata.procs[js_engine].poll() is not None \
+            or onWindows():
+        res = resource_stream(__name__, js_engine)
+        js_engine_code = res.read().decode('utf-8')
 
-    nodejs = localdata.proc
+        localdata.procs[js_engine] = new_js_proc(js_engine_code, force_docker_pull=force_docker_pull)
```
Comment (mr-c) [inline cwltool/sandboxjs.py]: Seems to be causing neverending loops on windows: https://ci.appveyor.com/project/mr-c/cwltool/build/.773-master/job/o57lcfpp9xjcrc7i#L1775

**Comment by mr-c** [comment]
Thanks again @ThomasHickman ; this is working everywhere but Windows. Ping me once that is fixed I'll do a final review and we can get this merged!

**Comment by ThomasHickman** [comment]
@mr-c I've fixed the windows issue. Do you want to do a final review now?

Note from above: I've removed the get_data to get_test_data change as I don't need `get_data` in `cwltool.utils` and this change affects a lot of files

**Comment by tetron** [comment]
The reason for using `resource_stream` is to support delivery as a Python egg.  Assuming files are on disk doesn't work when the module is stored in a zip archive (I don't know if wheels have similar behavior.)

**Comment by mr-c** [comment]
Yep, we know. But then it won't work if someone executes `cwltool.py`,
which is a shortcut for running cwltool without installing it first.

On Mar 12, 2018 15:22, "Peter Amstutz" <notifications@github.com> wrote:

The reason for using resource_stream is to support delivery as a Python
egg. Assuming files are on disk doesn't work when the module is stored in a
zip archive (I don't know if wheels have similar behavior.)

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/common-workflow-language/cwltool/pull/662#issuecomment-372326677>,
or mute the thread

**Comment by tetron** [comment]
So if we want to be clever about it, the wrapper method should try resource_stream() first and then fall back to searching for the file?

**Comment by ThomasHickman** [comment]
@tetron yes, this is already implemented in the util file in the tests folder: https://github.com/common-workflow-language/cwltool/blob/addcaa283cb76a035bd1c4741d4d4034194f242a/tests/util.py#L8. Also, if we are going to modify all `resource_stream` functions such that you can run code in place, wouldn't it be better to implement this in a separate PR?

**Comment by mr-c** [comment]
Sure, the decision that needs to be made can be moved to a separate issue
for discussion and (hopefully) quick implementation. @tetron, did you have
any other feedback on this PR?

**Comment by tetron** [comment]
A few comments - 

* Consider moving the call to JS validation out of load_tool and into `Process.__init__()`, that's where other secondary validation (input/output port type checking) currently happens.
* It looks like `get_expressions` traverse the schema and tool record in parallel.  This is a good strategy but somewhat expensive, a simpler/faster strategy would be to walk the schema once to determine the set of fields that are permitted to contain expressions, and then walk the tool object searching for those fields.
* If jshint finds an obvious error (syntax error, for example) it should raise a `ValidationException`

However I don't want to hold up merging a useful feature on these, if @mr-c has tested it and it seems to be working well, I'm okay with merging now and doing a followup cleanup PR.

**Comment by ThomasHickman** [comment]
@tetron Those seem fair enough comments. I also noticed while having a look at (1) that command line parameters were not getting propagated to validate_js in `load_tool` - the best way to solve this is to implement point (1), so this is done above

**Comment by mr-c** [comment]
@ThomasHickman @tetron Please file follow up issues as requested

**Comment by tetron** [comment]
@mr-c https://github.com/common-workflow-language/cwltool/issues/699 were there any others?

**Pull request merged**